Functional Code V1.0

gordonso928 opened the pull request:
- Code is functional with basic data logging
- Data logging need to be refined for post deploy analysis 
- algorithm( ) is tested with Logic_P_Control.cpp
- Teensy code is edited with PlatformIO, the file is .cpp instead if .ino
- Basic code review is needed

gordonso928 committed:
Initial

ToDo
-refine logic
-refine data writing method to SD card

gordonso928 committed:
Progress 15-Jul-2022 (0010)

- Refined yaw calculation function
- Started to change the overall structure of the code
- Will probably make some class for organized code

Commit:
15-Jul-2022 1259 Progress

gordonso928 committed:
Progress 16-Jul (0030)

- Started code tidy up
- Include info for the code and control flow
- tested and compiled with Arduino IDE
- Refined logic and read, save data process

gordonso928 committed:
Progress 16-Jul (1805)

Fixed syntax error

Commit:
Progress 18-Jul (1246)

- Started P Controller Routine

gordonso928 committed:
Progress 18-Jul (2327)

- Added logic test C++ folder
- Draft P control routine
- tested logic in .cpp, need to update logic in MAGPI_speedrun.ino
- checked program size, variables will need to recast

Commit:
[Progress] 19 Jul 1306

- Updated algorithm
- refine yaw integration

gordonso928 committed:
[Data Collection Update] 19-Jul 2101

data_collection.txt is created for @cristyro

gordonso928 committed:
19-Jul 2324 algorithm( ) Completed, Test Required

- Refined algorithms
- algorithm is able to handle any yaw angle (-inf, inf)
- logic tested with Logic_P_control.cpp
- Started to make reset routine
- Started to extract motor control routine for debug and control line winding code

Commit:
[Progress] 20-Jul 1254

- Further refined logic
- Started to prep for debug using #ifdef #endif

gordonso928 committed:
[20-Jul 2041] Version Save Before adding swRTC/Time

gordonso928 committed:
20-Jul 2229 Migrate to PlatformIO

Migrated project to PlatformIO for better debugging and coding

gordonso928 committed:
[21-Jul 0542] Debug Routine and CSV Structure

- Through out check of the code
- Refined elapsed time calculation with elapsedMicros or elapsedMillis
- Added Kalman filter for gyro data
- Added zero_motor.ino
- Added debug routines

Task left
- deployed condition
- refine data writing (e.g. writing data of the algorithm)
- start up (power mode of sensors)
- Raspberry PI Camera start up

gordonso928 commented:
- Motor control loop need to update
     - .read() will return the accumulated turn/position, no need to do turn-= motor.read

gordonso928 committed:
[JJ's review] 22 Jul 2120

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+#include "SimpleKalmanFilter.h"
+
+//states
+bool launched = 0;
+bool deployed = 0;
+bool test_status = 0;
```
Use false instead of 0, it is easier to read.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+#define EN 5 //replace pin number
+#define A1_3 3 //L293d channel 1A 3A 
+#define A2_4 4 //
+int motor_position = -999; // to be zero by the the hall effect sensor  
```
We are not using the hall sensor any more right? In that case this should start at 0.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+int motor_position = -999; // to be zero by the the hall effect sensor  
+Encoder motor(31, 32); 
+
+//algorithm
```
You should add some more comments for what each one of these are for and what units they have.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+  return angle;
+}
+
+void rotate_motor(double length){
```
This length is the absolute length the shroud is extended or retracted right? If so, leave a comment above the function that makes that clear.

gordonso928 replied:
My idea is you can pass in any length you need, but honestly that is more like a coordinate. Now I will just convert it to a "coordinate" of the motor with pos, and add the comment afterwards.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+}
+
+void rotate_motor(double length){
+  int pos = round(turn_per_length * length);
```
This will truncate automatically so you don't need the round()

gordonso928 replied:
I add round here as I know this is more like floor(), but since the resolution on the motor is high, I guess that does not matter much.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+  int pos = round(turn_per_length * length);
+  digitalWrite(EN, HIGH);
+  
+  if (pos > 0){
```
Add the updated P controller here.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+  
+}
+
+void algorithm(){
```
Leave a comment of what this algorithm actually does.

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+  yaw = 0.0;
+  elapsedMicros algorithm_time; //////////change back to elasped micros   //reset algorithm_time to find time passed and for integration
+
+  while (!(((error = yaw_req - yaw)<= yaw_control_cutoff)&&(error >= -yaw_control_cutoff)) && (algorithm_time < control_timeout)) {
```
I think this is equivalent right? All you're checking is if the error is small enough to stop.

```suggestion
  while (!(abs(error = yaw_req - yaw) <= yaw_control_cutoff)) && (algorithm_time < control_timeout)) {
```

gordonso928 replied:
I will change the code to this, totally forgot abs() exist

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+bool high_G(){
+  #ifdef REAL
+    mpu.getAccelerometerSensor()->getEvent(&a); //get only accleraiontion
+    if ((a.acceleration.x *a.acceleration.x + a.acceleration.y *a.acceleration.y + a.acceleration.z *a.acceleration.z) >= 25.0){
+      return 1;
+    } else {
+      return 0;
+    }
```
This is equivalent.

```suggestion
    return (a.acceleration.x *a.acceleration.x + a.acceleration.y *a.acceleration.y + a.acceleration.z *a.acceleration.z) >= 25.0;
```

gordonso928 replied:
This is much much better

jjderooy commented on MAGPI PlatformIO/src/main.cpp:
```diff
@@ -0,0 +1,461 @@
+  //add one set the motor to high impendace mode to prevent spinning
+  //add set sensor to sleep OR not active
+  
+  while(!launched){
+    void_timestamp = millis();
+    while(high_G() && !launched){  ///////////check if this is in Gs
+      launched = ((millis()- void_timestamp) > 200)? 1 : 0;
+    }
+  Serial.print("launched");
```
Leave a comment on the criteria for how it determines that it has launched.

jjderooy reviewed (commented):
Looks pretty good. Needs more commenting before I can do a full evaluation because there are parts that are difficult for me to understand.

gordonso928 committed:
[Progress] 23-Jul 0637

- Tested all sensors including GPS readings, all function properly with the code
- Updated code according to #4 PR
- Recoded motor_rotate(), note the logic should be correct but the actual output of the motor is unusual
- Added more comments for code review
- Added logic for state "deployed"

TODO
- PI camera code check (PI side)

gordonso928 committed:
Fix small mistake in 0637 progress

gordonso928 committed:
[progress] 25-Jul

gordonso928 committed:
30-Jul 2317

Updated file writing form main.cpp
New py file for py camera code

gordonso928 committed:
[Progress] 31-Jul 0319
- Comfirmed both cam working properly
- Resolved library issues
- Started on usb video capturing and recording

gordonso928 committed:
[Progress] 0430 31-Jul
- Updated logic to make picamera functional (at least one camera)
- Added teensy triggering condition
- Sat GPIO 8 as INPUT w/ PULLDOWN R (Normally Low)
- Have NOT look at combining the record file, will not have time to think a way, timestamp is used to create a unique file name

gordonso928 committed:
[PI Progress] 0502 31-Jul
- Compile and correct all syntax
- need to rewire the PI to MAGPI for better testing

Need to do
- chack and update activation (digitalWrite()) on teensy
- check if the code work
- check if we have time to complete usb camera tommorow

gordonso928 commented:
Close for new PR
